lua/wikis/commons/CharacterStats.lua

```diff
@@ -0,0 +1,242 @@
+---@return string[]
+function CharacterStats:getTeamBans(game, opponentIndex)
+	return Array.filter(Array.mapIndexes(function (characterIndex)
+		return String.nilIfEmpty(game.extradata['team' .. opponentIndex .. 'ban' .. characterIndex])
+	end), String.isNotEmpty)
+end
+
+---@param game CharacterStatsGame
+---@param opponentIndex integer
+---@return string?
+function CharacterStats:getTeamSide(game, opponentIndex)
+	return String.nilIfEmpty(game.extradata['team' .. opponentIndex .. 'side'])
+end
+
+---@return string[]
+function CharacterStats:getSides()
+	error('CharacterStats:getSides() cannot be called directly and must be overridden.')
+end
+
+---@param games CharacterStatsGame[]
+---@return {characterData: CharacterStatistic[], overall: {games: integer, wins: table<string, integer>}}
+function CharacterStats:processGames(games)
+	---@type table<string, CharacterStatistic>
+	local stats = {}
+
+	local overallData = {
+		games = #games,
+		wins = Table.map(self:getSides(), function (key, value)
+			return value, 0
+		end)
+	}
+
+	Array.forEach(games, function (game)
+		local teamsCharacters = {
+			self:getTeamCharacters(game, 1),
+			self:getTeamCharacters(game, 2)
+		}
+		Array.forEach(game.matchOpponents, function (opponent, opponentIndex)
+			local opponentName = Opponent.toName(opponent)
+			local isWinner = game.winner == opponentIndex
+			local side = self:getTeamSide(game, opponentIndex)
+			local sideKey = isWinner and 'win' or 'loss'
+
+			if isWinner then
+				overallData.wins[side] = overallData.wins[side] + 1
+			end
+
+			local characters = teamsCharacters[opponentIndex]
+
+			Array.forEach(characters, function (character, characterIndex)
+				local characterStats = self:_getCharacterStatTable(stats, character)
+				if not characterStats.playedBy[opponentName] then
+					characterStats.playedBy[opponentName] = {pick = 0, win = 0, loss = 0}
+				end
+				characterStats.total.pick = characterStats.total.pick + 1
+				characterStats.playedBy[opponentName].pick = characterStats.playedBy[opponentName].pick + 1
+
+				characterStats.total[sideKey] = characterStats.total[sideKey] + 1
+				characterStats.side[side][sideKey] = characterStats.side[side][sideKey] + 1
+				characterStats.playedBy[opponentName][sideKey] = characterStats.playedBy[opponentName][sideKey] + 1
+
+
+				Array.forEach(characters, function (playedWithCharacter, playedWithCharacterIndex)
+					if characterIndex == playedWithCharacterIndex then
+						return
+					elseif not characterStats.playedWith[playedWithCharacter] then
+						characterStats.playedWith[playedWithCharacter] = {pick = 0, win = 0, loss = 0}
+					end
+					characterStats.playedWith[playedWithCharacter].pick =
+						characterStats.playedWith[playedWithCharacter].pick + 1
+					characterStats.playedWith[playedWithCharacter][sideKey] =
+						characterStats.playedWith[playedWithCharacter][sideKey] + 1
+				end)
+
+				Array.forEach(teamsCharacters[3 - opponentIndex], function (playedAgainstCharacter)
```

Comment:
this assumes we have exactly 2 opponents
do we check for that anywhere?
(reviewing on mobile so kinda shit to check properly)

Reply:
i think warcraft might be interested in it
(they have heroes per map too)
and they potentially might get ffa too